### PR TITLE
:bug: enable osv-scanner Go call analysis

### LINKED ISSUE #3891: scorecard started reducing score for vulnerabilities in unrelated packages that aren't imported
**Describe the bug**

Scorecard used to show 0 vulnerabilities but it suddenly started showing 29 vulnerabilities (e.g. in packages that are not imported directly or indirectly).  This reduces the score by 10 points under "Vulnerabilities".

For example, a [CBOR codec](https://github.com/fxamacker/cbor) library isn't using any networking packages.  However, scorecard is showing it as being affected by issues in`net/http` such as GO-2022-0969.

> Warn: Project is vulnerable to: GO-2022-0969

[GO-2022-0969](https://pkg.go.dev/vuln/GO-2022-0969) is a bug in `net/http`, which isn't directly or indirectly imported by the CBOR codec.

**Reproduction steps**

1.  view score at https://securityscorecards.dev/viewer/?uri=github.com/fxamacker/cbor/v2
2.  expand "Vulnerabilities" (it shows 29 vulnerabilities like GO-2022-0969 which is in `net/http`)
3. view imports at https://pkg.go.dev/github.com/fxamacker/cbor/v2?tab=imports
4. notice that `net/http` is not imported by fxamacker/cbor or fxamacker/cbor/v2
 
**Expected behavior**

Report 0 vulnerabilities from `net/http` for projects that don't use or import anything that uses `net/http`.

**Additional context**

Screenshot

image

I didn't investigate remaining vulnerabilities after seeing the first one is a false alarm.

### PR DESCRIPTION
**Pull request by spencerschrock**
#### What kind of change does this PR introduce?

bug fix

- [X] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?
The `Vulnerabilities` check deducts for Go vulnerabilities which aren't called.

#### What is the new behavior (if this is a feature change)?**
`osv-scanner` Go call analysis is enabled. Uncalled vulnerable functions aren't included in the build, and don't affect the score.

- [ ] Tests for the changes have been added (for bug fixes/features)

I'm not sure there's a great way of testing this via unit tests at least. Perhaps an e2e test?
#### Which issue(s) this PR fixes
Fixes #3891 

#### Special notes for your reviewer

Testing locally on the original repo seems to have resolved the issue.
```shell
$ go run main.go --repo github.com/fxamacker/cbor --checks Vulnerabilities --show-details --format json | jq
{
  "date": "2024-02-20T15:24:30-08:00",
  "repo": {
    "name": "github.com/fxamacker/cbor",
    "commit": "afbafc4054c62138633ff49dfcded790b11e44de"
  },
  "scorecard": {
    "version": "",
    "commit": "unknown"
  },
  "score": 10.0,
  "checks": [
    {
      "details": null,
      "score": 10,
      "reason": "0 existing vulnerabilities detected",
      "name": "Vulnerabilities",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/main/docs/checks.md#vulnerabilities",
        "short": "Determines if the project has open, known unfixed vulnerabilities."
      }
    }
  ],
  "metadata": null
}
```
#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
The Vulnerabilities check now ignores uncalled Go vulnerabilities.
```

### CODE REVIEW

**Comment by spencerschrock** [comment]
@hogo6002 would you mind taking a look at this? I couldn't find a way of asking `osvscanner.DoScan` for all stable call analyses, so for now I just enabled Go explicitly. 

Additionally in the commit message I claim "uncalled vulnerable functions aren't included in the build", which I think is true for Go, but figured someone on OSV would know better.

**Review by naveensrinivasan — APPROVED** [review]

**Comment by spencerschrock** [comment]
I'm trying to create a local test. A project which uses an old version of Go intentionally.
Based on my understanding of https://osv.dev/vulnerability/GO-2023-2102:
`http.ListenAndServe` is vulnerable for versions before 1.20.10 (or 1.21.3)
```go
package main

import (
    "log"
    "net/http"
)

func main() {
    err := http.ListenAndServe(":8080", nil)
    if err != nil {
        log.Fatal(err)
    }
}
```
the example `go.mod` declares a vulnerable version:
```
module example.com/foo/bar

go 1.19
```

But `osv-scanner` seems to also care what my Go version is when determining reachability (go.mod still says 1.19).
```shell
$ go version
go version go1.22.0 linux/amd64

$ osv-scanner -r .
Scanning dir .
Scanned /tmp/go/go.mod file and found 1 package
Uncalled vulnerabilities 
│ https://osv.dev/GO-2023-2375 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2041 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2043 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2102 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2185 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2186 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2382 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
```
Vs if I use an older Go toolchain (example go.mod is still 1.19).

```shell
$ go version
go version go1.21.2 linux/amd64
$ osv-scanner -r .
Scanning dir .
Scanned /tmp/go/go.mod file and found 1 package
(called vulnerabilities section)
│ https://osv.dev/GO-2023-2102 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2185 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2382 │      │ Go        │ stdlib  │ 1.19    │ go.mod │

Uncalled vulnerabilities
│ https://osv.dev/GO-2023-2375 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2041 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2043 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
│ https://osv.dev/GO-2023-2186 │      │ Go        │ stdlib  │ 1.19    │ go.mod │
```

Just trying to understand this behavior before enabling it. As our weekly scan runs with the latest version of the Go toolchain

File: clients/osv.go [hunk 1]
```diff
@@ -52,6 +52,9 @@ func (v osvClient) ListUnfixedVulnerabilities(
 		SkipGit:        true,
 		Recursive:      true,
 		GitCommits:     gitCommits,
+		CallAnalysisStates: map[string]bool{
+			"go": true,
+		},
```
Comment (spencerschrock) [inline clients/osv.go:57]: putting a temporary merge hold while discussing exactly how this would work

**Comment by hogo6002** [comment]
> But osv-scanner seems to also care what my Go version is when determining reachability (go.mod still says 1.19).

@spencerschrock Thanks for reporting this issue. We just merged a fix PR (https://github.com/google/osv-scanner/pull/818) for it. 

The installed Go version does not matter to osv-scanner when generating the reachability results. 
But be mindful that we might assume a vulnerability has been called if it doesn't have specific function-level details. This information is usually missing from GHSAs unless their aliases have it.

**Comment by spencerschrock** [comment]
> > But osv-scanner seems to also care what my Go version is when determining reachability (go.mod still says 1.19).
> 
> @spencerschrock Thanks for reporting this issue. We just merged a fix PR ([google/osv-scanner#818](https://github.com/google/osv-scanner/pull/818)) for it.

Is osv-scanner planning on a patch release, or should Scorecard consider doing a pseudo-version update to pull the patch in?

> The installed Go version does not matter to osv-scanner when generating the reachability results. But be mindful that we might assume a vulnerability has been called if it doesn't have specific function-level details. This information is usually missing from GHSAs unless their aliases have it.

Scorecard currently assumes all reported OSV vulns are called, so I think that would be identical to how things are now.

**Comment by hogo6002** [comment]
> Is osv-scanner planning on a patch release, or should Scorecard consider doing a pseudo-version update to pull the patch in?
Yes, we will do a release this week.

**Comment by spencerschrock** [comment]
So this works locally, but when run as a docker image (which is what the cron and scorecard action will do), enabling call analysis doesn't actually do anything because the docker image only has the scorecard binary, not a copy of the Go binary

**Pull request closed without merging**